CIP-0037/CIP-0037.md

```diff
@@ -0,0 +1,352 @@
+others)
+
+
+### Reason for choosing BLS12_381
+The reason for choosing BLS12_381 over BN256 curve is that the former is claimed to provide 128 bits of security, 
+while the latter was reduced to 100 bits of security after the extended number field sieve (a new algorithm to compute 
+the discrete logarithm) was [shown to reduce the security](https://eprint.iacr.org/2016/1102.pdf) of these curves. 
+
+An [EIP](https://eips.ethereum.org/EIPS/eip-2537) for precompiles of curve BLS12_381 already exists, but has been 
+stagnant for a while. Nonetheless, Zcash, MatterLabs and Consensys support BLS12_381 curve, so it is certainly widely 
+used in the space. 
+
+Further reading regarding curve BLS12_381 can be found [here](https://hackmd.io/@benjaminion/bls12-381) and the 
+references thereof cited. For level of detail required in this document, it is sufficient to understand that one can 
+define a pairing over BLS12_381, where a pairing is a map: e:G1 X G2 -> GT, which satisfies the following properties:
+
+* Bilinearity: for all a,b in F^*_q, for all P in G1, Q in G2: e(aP,bQ)=e(P,Q)^(ab)
+* Non-degeneracy: e != 1
+* Computability: There exists an efficient algorithm to compute e
+
+where G1, G2 and GT are three distinct groups of order a prime q. Given that all three groups have the same 
+order, we can refer to the scalars of each group using the same type/structure. 
+
+## Specification
+We now provide the technical specification. 
+
+### Names and types/kinds for the new functions or types
+The added types will be the following, all of which can be represented as a byte array. Even if these types are 
+equivalent to byte arrays of a given size, it makes sense to include these types, to enforce deserialization, and 
+therefore some checks on the data used by the smart contract. In particular, all three types can only be achieved 
+with byte arrays that represent a point which is part of the prime order subgroup. 
+
+* `BLS12_381G1Element`
+* `BLS12_381G2Element`
+* `BLS12_381GTElement`
+
+We need to support the binary operation of G1 and G2 (which are additive groups), as well as the binary operation 
+over GT (which is represented as a multiplicative group). We also want to enable hashing to G1 and G2.
+In particular, we expose the hash to curve (which we denote with `H2G1` and `H2G2` for G1 and G2 
+respectively) algorithm as described in 
+[hash to curve draft](https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-hash-to-curve#section-8), 
+using `BLS12381G1_XMD:SHA-256_SSWU_RO_` and `BLS12381G2_XMD:SHA-256_SSWU_RO_` for G1 and G2 respectively.
+We do not include the type of the Scalar, nor operations related to it. These type of operations can already be 
+performed with the Integral type. In particular, we need the following functions: 
+
+* Group operations: 
+    * `BLS12_381_G1_add :: BLS12_381G1Element -> BLS12_381G1Element -> BLS12_381G1Element`
```

Comment:
Idea that @morganthomas told me, we could have each of these generic group operations be indexed with the type of the group essentially, so `add :: (g :: G) -> g -> g -> g` in a dependently typed language, where `g` is internally represented as an integer. In the implementation, the interpreter would then err for invalid choices of `g`.

Reply:
This might be better for supporting other curves in the future, what do you think @michaelpj ?

Reply:
I think having a generic strategy is worth considering, though we need the opinion of @michaelpj here. I don't think we will add support for a lot of pairing friendly curves though, as this moves somewhat slowly. But in the mid/long term, this might be the case. 

My understanding is that the motivation for this is to include non-pairing friendly curves (such as pasta curves). My concern is that this generic strategy will become quite complex if we want to support all kind of EC families, and I'm not sure I see a huge benefit over simply adding support for different curves. For instance, do we want the interpreter to be aware of the cycles of the pasta curves (or any cycles of pairing friendly curves)? how would the pairing fit in such a generic framework? Would we have something like `BLS12_381_pairing :: (g1 :: G, g2::G) -> g1 -> g2 -> something`? (sorry if my notation is incorrect, I'm not a haskell developer). How would we represent the target group? 

I think that having a generic strategy that supports pairing friendly and non-pairing friendly curves would be beneficial, as we could also use it to expose curve operations over secp256k1 or curve25519.

Reply:
I don't think we can do this easily. Firstly, we don't have the dependent typing. Secondly, if we added new curves to the support here, that would mean that some builtin calls which had previously failed would start succeeding, which is a visible semantics change.

Reply:
@michaelpj I agree in that it *is* a change to the semantics, but isn't that true for *every* change? A transaction that attempts to use Plutus V2 will fail now, but won't in the future, no?

wrt. dependent typing, that was simply to make it clear what the intent was. You could have the signature be `add :: Int -> G -> G -> G` where `G` is a sum type that represents all the possible choices.

Reply:
@michaelpj I agree in that it *is* a change to the semantics, but isn't that true for *every* change? A transaction that attempts to use Plutus V2 will fail now, but won't in the future, no?

wrt. dependent typing, that was simply to make it clear what the intent was. You could have the signature be `add :: Int -> G -> G -> G` where `G` is a sum type that represents all the possible choices.

Reply:
It's a change to the semantics of *existing scripts*. That we can't do easily. And sadly even failures are visible, because you can post a failed script to the chain if it's a phase 2 failure.

Reply:
@michaelpj Well the Plutus version could determine what indices are valid. That would solve the issue AFAICT.

Reply:
If you were willing to do a new Plutus version every time. But we're expecting those to be very infrequent, we can normally add builtins without needing one.

Reply:
Well that sounds fine to me honestly. We don't want that many anyway.
However, can you think of a better solution rather than this? Is it more sensible to duplicate the operations for each curve? Are we duplicating type checking functionality? What is the most efficient? Is having that many builtins (O(n) for n curves) practical? Could the implementation be simpler if they were the same builtin?

Reply:
They would certainly not have identical performance characteristics. And as I said above, different curves (or more precisely, groups) will even have different functions. For instance, if we were to include operations over the group defined over secp256k1, we would only need `add`, `mult` and `neg` functions once (together with serialization and equality checks), but there would be no concept of pairing, `G1` or `G2`, and target group. If, on the other hand, we were to include operations over the groups defined over the pasta curves, we would need those operations together with some definition of `cycles` (something not covered in this CIP). I'm still not sure that having a single generic API covering all types of groups over ECs (including non-prime order groups, pairing friendly curves and cycles over different curves) will be simpler, provided that we don't plan on supporting a very big number of curves.

Reply:
I agree with this, albeit it's unfortunate.